src/amd64/amd64-51-30k-choose_tp.s

```diff
@@ -0,0 +1,2189 @@
+movq   72(%rcx,%rdi),%rbp
+
+# qhasm: txaddy4 = t if =
+# asm 1: cmove <t=int64#15,<txaddy4=int64#14
+# asm 2: cmove <t=%rbp,<txaddy4=%rbx
+cmove %rbp,%rbx
+
+# qhasm: =? u - 2
+# asm 1: cmp  $2,<u=int64#5
+# asm 2: cmp  $2,<u=%r8
+cmp  $2,%r8
+
+# qhasm: t = *(uint64 *)(basep + 160 + pos)
+# asm 1: movq   160(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   160(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   160(%rcx,%rdi),%rbp
+
+# qhasm: tysubx0 = t if =
+# asm 1: cmove <t=int64#15,<tysubx0=int64#2
+# asm 2: cmove <t=%rbp,<tysubx0=%rsi
+cmove %rbp,%rsi
+
+# qhasm: t = *(uint64 *)(basep + 168 + pos)
+# asm 1: movq   168(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   168(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   168(%rcx,%rdi),%rbp
+
+# qhasm: tysubx1 = t if =
+# asm 1: cmove <t=int64#15,<tysubx1=int64#6
+# asm 2: cmove <t=%rbp,<tysubx1=%r9
+cmove %rbp,%r9
+
+# qhasm: t = *(uint64 *)(basep + 176 + pos)
+# asm 1: movq   176(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   176(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   176(%rcx,%rdi),%rbp
+
+# qhasm: tysubx2 = t if =
+# asm 1: cmove <t=int64#15,<tysubx2=int64#7
+# asm 2: cmove <t=%rbp,<tysubx2=%rax
+cmove %rbp,%rax
+
+# qhasm: t = *(uint64 *)(basep + 184 + pos)
+# asm 1: movq   184(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   184(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   184(%rcx,%rdi),%rbp
+
+# qhasm: tysubx3 = t if =
+# asm 1: cmove <t=int64#15,<tysubx3=int64#8
+# asm 2: cmove <t=%rbp,<tysubx3=%r10
+cmove %rbp,%r10
+
+# qhasm: t = *(uint64 *)(basep + 192 + pos)
+# asm 1: movq   192(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   192(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   192(%rcx,%rdi),%rbp
+
+# qhasm: tysubx4 = t if =
+# asm 1: cmove <t=int64#15,<tysubx4=int64#9
+# asm 2: cmove <t=%rbp,<tysubx4=%r11
+cmove %rbp,%r11
+
+# qhasm: t = *(uint64 *)(basep + 200 + pos)
+# asm 1: movq   200(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   200(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   200(%rcx,%rdi),%rbp
+
+# qhasm: txaddy0 = t if =
+# asm 1: cmove <t=int64#15,<txaddy0=int64#10
+# asm 2: cmove <t=%rbp,<txaddy0=%r12
+cmove %rbp,%r12
+
+# qhasm: t = *(uint64 *)(basep + 208 + pos)
+# asm 1: movq   208(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   208(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   208(%rcx,%rdi),%rbp
+
+# qhasm: txaddy1 = t if =
+# asm 1: cmove <t=int64#15,<txaddy1=int64#11
+# asm 2: cmove <t=%rbp,<txaddy1=%r13
+cmove %rbp,%r13
+
+# qhasm: t = *(uint64 *)(basep + 216 + pos)
+# asm 1: movq   216(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   216(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   216(%rcx,%rdi),%rbp
+
+# qhasm: txaddy2 = t if =
+# asm 1: cmove <t=int64#15,<txaddy2=int64#12
+# asm 2: cmove <t=%rbp,<txaddy2=%r14
+cmove %rbp,%r14
+
+# qhasm: t = *(uint64 *)(basep + 184 + pos)
+# asm 1: movq   184(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   184(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   224(%rcx,%rdi),%rbp
```

Comment:
Comparing to the original, the changes up till now are just subtracting 40. On this one, the comments points to the "new" offset, though the code matches the expected progression. Why is that ? Simple bug in the original ? If so, bug in the comment, or the actual code ?

Reply:
The bug is almost certainly in the comment. The wallet PR has a benchmark utility which first runs a diagnostic check ensuring that the existing output from the `crypto::...` functions is identical to these accelerated versions. An offset bug here would be multiplying the secret key bits incorrectly, and the probability of it matching repeatedly are pretty small.

I will double check to see what happened, most likely I just forgot to update the comment. Initially I wasn't updating the comments at all, but part way through I started updating them. Must've missed some on the re-pass back.